File: src/utils/api/reportCounts.ts
import API_ENDPOINTS from "src/config/api";

interface CountsDetails {
  counts: number;
  reasons?: {
    [key: string]: number;
  };
}

interface StatusCounts {
  [key: string]: CountsDetails;
  delivered: CountsDetails;
  failed: CountsDetails;
  processing: CountsDetails;
}

export interface ReportCounts {
  total_counts: number;
  status_counts: StatusCounts;
}

export const defaultReportCounts: ReportCounts = {
  total_counts: 0,
  status_counts: {
    delivered: { counts: 0 },
    failed: { counts: 0, reasons: {} },
    processing: { counts: 0 },
  },
};

const getReportCounts = async (
  access_token: string,
  data_stream_id: string,
  data_stream_route?: string,
  date_start?: string,
  date_end?: string,
  days_interval?: number
): Promise<Response> => {
  const params = new URLSearchParams();

  if (data_stream_id) params.append("data_stream_id", data_stream_id);
  if (data_stream_route) params.append("data_stream_route", data_stream_route);
  if (date_start) params.append("date_start", date_start);
  if (date_end) params.append("date_end", date_end);
  if (days_interval) params.append("days_interval", days_interval.toString());

  const url = `${API_ENDPOINTS.reportCounts}?${params.toString()}`;

  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: "Bearer " + access_token,
    },
  }).catch();

  return response;
};

export default getReportCounts;

File: api/src/main/kotlin/dexportal/routes/psApi/ReportCounts.kt
package dexportal.routes.psApi

import com.apollographql.apollo3.ApolloClient

import dexportal.config.ConfigLoader
import dexportal.utils.processDateString
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*

import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import io.ktor.server.routing.*

import com.apollographql.apollo3.api.toJson
import com.dexportal.SampleQuery

fun Route.reportCounts(apolloClient: ApolloClient) {
    route("/report-counts") {
        get {
            // Make request to GraphQL endpoint, return it as json
            val response = apolloClient.query(SampleQuery()).execute()
            val jsonData = response.data?.toJson()
            call.respond(HttpStatusCode.OK, jsonData.toString())

//            call.respond(HttpStatusCode.OK, "This is the /report-counts endpoint")
        }
    }
}
// fun Route.reportCounts(client: HttpClient) {
//     val psApiUrl = ConfigLoader.getPsApiEndpoint()
//
//     route("/report-counts") {
//         get {
//             val dataStreamId = call.request.queryParameters["data_stream_id"] ?: ""
//             val dataStreamRoute = call.request.queryParameters["data_stream_route"] ?: ""
//             val dateStart = call.request.queryParameters["date_start"] ?: ""
//             val dateEnd = call.request.queryParameters["date_end"] ?: ""
//             val daysInterval = call.request.queryParameters["days_interval"] ?: ""
//
//             val authToken = call.request.headers["Authorization"]
//
//             val cleanedDateStart = processDateString(dateStart) ?: ""
//             val cleanedDateEnd = processDateString(dateEnd) ?: ""
//
//             try {
//                 val response: HttpResponse = client.get("$psApiUrl/api/report/counts/submissions/summary") {
//                     header("Authorization", authToken)
//                     parameter("data_stream_id", dataStreamId)
//                     parameter("data_stream_route", dataStreamRoute)
//                     if (cleanedDateStart.isNotEmpty()) {
//                         parameter("date_start", cleanedDateStart)
//                     }
//                     if (cleanedDateEnd.isNotEmpty()) {
//                         parameter("date_end", cleanedDateEnd)
//                     }
//                     if (daysInterval.isNotEmpty()) {
//                         parameter("days_interval", daysInterval)
//                     }
//                 }
//                 val responseBody: String = response.bodyAsText()
//                 call.respondText(responseBody)
//             } catch (e: Exception) {
//                 call.respond(HttpStatusCode.InternalServerError, e.message ?: "An unknown error occurred")
//             }
//         }
//     }
// }
